### PR TITLE
RecordType attribute

### PR DESCRIPTION
**Pull request by nielsdisveld**
In this PR:
- Migrator logic now has its own class
- RecordTypeCache provides cached dictionaries to find correct Record.Type corresponding to Record.GetType() and vice versa.
- RecordTypeCache contains a static dictionary based on types found in entire assembly for when it has not yet been initialized
- RecordTypeAttribute for when a custom Record.Type is needed (different from Record.GetType().Name)
- Simple tests for custom record types

### CODE REVIEW

**Comment by nielsdisveld** [comment]
TODO: Discuss if we want to use Reflection this much

File: EventSourcing.Core/Migrators/RecordMigratorService.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        _migrators = (options?.MigratorTypes ?? AssemblyMigratorTypes)
+            .Select(type => Activator.CreateInstance(type) as IRecordMigrator)
+            .ToDictionary(migrator => migrator!.Source, migrator => migrator);
+        
+        ValidateMigrators();
+    }
+    
+    public Record Migrate(Record record)
+    {
+        while (_migrators.TryGetValue(record.GetType(), out var migrator))
```
Comment (BramKraai) [inline EventSourcing.Core/Migrators/RecordMigratorService.cs:25]: I like this while loop a lot! :)

File: EventSourcing.Core/Records/RecordTypesCache.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
+using System.Reflection;
+
+namespace EventSourcing.Core.Types;
+
+public sealed class RecordTypesCache
```
Comment (BramKraai) [inline EventSourcing.Core/Records/RecordTypesCache.cs:5]: Maybe rename to ```RecordTypeCache``` ?

File: EventSourcing.Core/Records/RecordTypesCache.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    private static readonly List<Type> AssemblyRecordTypes = AppDomain.CurrentDomain
+        .GetAssemblies()
+        .SelectMany(x => x.GetTypes())
+        .Where(type => typeof(Record).IsAssignableFrom(type) && type.IsClass && !type.IsAbstract && type.IsPublic)
+        .ToList();
+    private static readonly Dictionary<Type, string> RecordTypesRev;
+    static RecordTypesCache()
+    {
+        // Create dictionaries mapping from Record.Type to Record Type string
+        RecordTypesRev = AssemblyRecordTypes.ToDictionary(type => type, type => type.GetCustomAttribute<RecordType>()?.Value ?? type.Name);
```
Comment (BramKraai) [inline EventSourcing.Core/Records/RecordTypesCache.cs:17]: The ```RecordTypesRev``` definition and initialisation can be merged. The static Constructor will then not be needed :)

File: EventSourcing.Cosmos.Tests/CosmosRecordAttributeTests.cs [hunk 4]
```diff
@@ -0,0 +1,30 @@
+using EventSourcing.Core;
+using EventSourcing.Core.Tests;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Options;
+
+namespace EventSourcing.Cosmos.Tests;
+
+public class CosmosRecordAttributeTests: RecordAttributeTests
```
Comment (BramKraai) [inline EventSourcing.Cosmos.Tests/CosmosRecordAttributeTests.cs:8]: Downside of having these different tests classes in the core file is that the Cosmos configuration and initialisation needs to be done multiple times. It would be nicer to have one tests suite class that a particular implementation has to implement. Kind of like the ```EventStoreTests```, where I used a partial class to split one class up in multiple files.

File: EventSourcing.Core/Migrators/RecordMigratorService.cs [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+{
+    private static List<Type> AssemblyMigratorTypes => AppDomain.CurrentDomain
+        .GetAssemblies()
+        .SelectMany(assembly => assembly.GetTypes())
+        .Where(type => typeof(IRecordMigrator).IsAssignableFrom(type) && type.IsClass && !type.IsAbstract && type.IsPublic)
+        .ToList();
+    
+    private readonly Dictionary<Type, IRecordMigrator?> _migrators;
+    
+    public RecordMigratorService(RecordConverterOptions? options = null)
```
Comment (BramKraai) [inline EventSourcing.Core/Migrators/RecordMigratorService.cs:13]: This should probably just accept a list of Migrator Types

File: EventSourcing.Core/Records/Record.cs [hunk 6]
```diff
@@ -48,7 +51,7 @@ public record Record
   protected Record()
   {
     RecordId = Guid.NewGuid();
-    Type = GetType().Name;
+    Type = RecordTypeCache.GetRecordTypeStringStatic(GetType());
```
Comment (BramKraai) [inline EventSourcing.Core/Records/Record.cs:54]: Maybe rename to ```.GetAssemblyRecordTypeString``` ?

File: EventSourcing.Core/Records/RecordTypeCache.cs [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    }
+    public Type GetRecordType(string typeString)
+    {
+        if (!_recordTypes.TryGetValue(typeString, out var type))
+            throw new InvalidOperationException($"Record with Type '{typeString}' not found");
+
+        return type;
+    }
+    public string GetRecordTypeString(Type type) => GetRecordTypeStringStatic(type, _recordTypeStrings);
+    public static string GetRecordTypeStringStatic(Type type, Dictionary<Type, string>? recordTypeStrings = null)
```
Comment (BramKraai) [inline EventSourcing.Core/Records/RecordTypeCache.cs:39]: Maybe rename to ```GetAssemblyRecordTypeString``` ?

File: EventSourcing.Core/Records/RecordValidation.cs [hunk 8]
```diff
@@ -14,42 +17,44 @@ public static void ValidateEventSequence(Guid partitionId, IList<Event> events)
     foreach (var e in events)
       ValidateRecord(e);
 
-    const string message = "Error Validating Event Sequence. ";
+    const string message = "Error Validating Event Sequence: ";
 
     var partitionIds = events.Select(x => x.PartitionId).Distinct().ToList();
 
     if (partitionIds.Count > 1)
-      throw new RecordValidationException(message + $"All Events must share the same PartitionId. Expected {partitionId}, Found [ {string.Join(", ", events.Select(x => x.PartitionId))} ]");
```
Comment (BramKraai) [inline EventSourcing.Core/Records/RecordValidation.cs:22]: Why did you remove the long form Exceptions?
